### DISCUSSION TITLE
Sandbox redflags

### DISCUSSION THREAD

**Discussion by rusty-snake** [netblue30/firejail#5503] [General]
List of redflags for sandboxes indicating a *very* weak sandbox that can be escaped easily i.e. not a true sandbox. And the firejail commands to block those access.

- D-Bus access
  - `dbus-{user,system} (filter|none)` [^1]
- Unrestricted `/dev` access
  - `private-dev`
- Network access
  - `net none`
  - `seccomp.drop socket`
- Write access to `${HOME}`
  - `private`
  - `private ${HOME}/foo`
  - `whitelist ${HOME}/foo`
  - `private-home foo,bar`
  - `read-only ${HOME}`
- X11 access
  - `x11 none`
  - `blacklist /tmp/.X11-unix`

[^1]: If using `filter` there must be no dangerous `own`/`talk` rules.

**Comment by rusty-snake** [comment]
When we define sandboxing as drawing a (hard) line between a trusted domain and a untrusted domain (an alternative could be CIA Model), we can group sandboxing steps into following categories:

1. Disables sandbox escape technology (i.e. `whitelist ${HOME}/foo`)
   - The steps in this group disable everything that can be used to escape the sandbox by the use of features (!=bugs).
   - It is the most important group.
   - If you fail here it is a sandbox escape vulnerability.
   - If you - for any reason - can not *fully* implement *all* steps in this group you can skip the rest, everything else would be security theater.
2. Kernel and Userspace Attack surface reduction, limitations of available OS features, limitation of accessible data about the system ... in short **defense-in-depth** (i.e. `nonewprivs` or `seccomp` <sub>plays also a role in sandbox escape technology</sub>)
   - The steps in this group disable everything that could be used to escape the sandbox by the use of (zero day) bugs (!=features) and isn't needed by the sandboxed program.
3. Limitation of accessible data and files containing sensitive (i.e. `blacklist ${HOME}/.gnupg`)
   - Not a goal for us but I don't wanted it in the next group because it is a goal when using CIA.
4. Mitigations to approximate 1. if it is not possible (i.e. `blacklist ${HOME}/.bashrc`)
   - Security theather because you failed in 1. and no longer draw a hard, solid line.
   - I sometimes call this "save your ass security" because it can stop malicious code by providing an unexpected environment without really stopping the attackers possibilities.

**Comment by kmk3** [comment]
@rusty-snake [on Dec 3](https://github.com/netblue30/firejail/discussions/5503#discussion-4624875):

> List of redflags for sandboxes indicating a _very_ weak sandbox that can be
> escaped easily i.e. not a true sandbox.

Given the contents of the list, I assume you mean that missing any of those
entries is a redflag in a given profile, not that their presence is a redflag.

Rewording suggestion:

> List of very important security-related entries for sandboxes.  Missing any
> of them in a profile is a redflag that indicates a _very_ weak sandbox that
> can be escaped easily (i.e. not a true sandbox).

**Reply by rusty-snake** [reply]
The outer list lists the redflags and the inner lists list the firejail options to block those accesses.

**Reply by kmk3** [reply]
@rusty-snake [on Dec 4](https://github.com/netblue30/firejail/discussions/5503#discussioncomment-4302375):

> The outer list lists the redflags and the inner lists list the firejail
> options to block those accesses.

I see it now, though if you hadn't said it I would probably never have guessed
it, as the commands in the inner lists do basically the opposite of what their
respective parent item says (such as with "`private-dev`" and "Unrestricted
`/dev` access").

I think that being a bit more explicit would make it much clearer.

For example:

> List of key restrictions for sandboxes.  Missing any of them in a profile is
> a redflag that indicates a _very_ weak sandbox that can be escaped easily
> (i.e. not a true sandbox).
> 
> Block/limit D-Bus access:
> 
> * `dbus-user (filter|none)` [1](#user-content-fn-1-d8261cbf66d0def5832f799cfac4f3ee)
> * `dbus-system (filter|none)` [1](#user-content-fn-1-d8261cbf66d0def5832f799cfac4f3ee)
> 
> Limit `/dev` access:
> 
> * `private-dev`
> 
> Block/limit network access:
> 
> * `net none`
> * `seccomp.drop socket`
> 
> Block/limit write access to (the real) `${HOME}`:
> 
> * `private`
> * `private ${HOME}/foo`
> * `whitelist ${HOME}/foo`
> * `private-home foo,bar`
> * `read-only ${HOME}`
> 
> Block X11 access:
> 
> * `x11 none`
> * `blacklist /tmp/.X11-unix`

**Comment by topimiettinen** [comment]
There could be a tool like `systemd-analyze security`, which would calculate a score for all profiles, or for one specific in more detail.

**Reply by rusty-snake** [reply]
I wrote [raudit](https://github.com/rusty-snake/raudit) which calculates a score based on the sandbox (unlike its definition). But it has only three implemented rules.
